### ISSUE TITLE
[Bug]: Major subprocess leak in WebKitGTK applications

### ISSUE THREAD

**Issue by mcatanzaro** [flatpak/flatpak#5946]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.10

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

41

### What architecture are you using?

x86_64

### How to reproduce

Unknown -- happens very commonly in regular usage, but never when I try to figure out how to reproduce

### Expected Behavior

The intermediate processes (flatpak-spawn, bwrap, and xdg-dbus-proxy) should exit when the WebKit subprocess exits

### Actual Behavior

The intermediate processes outlive the corresponding WebKit subprocess and just keep running forever, even after the application is closed

### Additional Information

Originally reported as https://bugs.webkit.org/show_bug.cgi?id=279913

Inspecting flatpak-spawn.c, I think the only way this happens is if the flatpak portal doesn't emit the SpawnExited D-Bus signal. (Or if flatpak-spawn is somehow broken and doesn't receive it for some reason.) Since xdg-dbus-proxy is still running as well, I *assume* the flatpak portal has failed to notice the WebKit subprocess exiting.

**Comment by mcatanzaro** [comment]
Unfortunately I don't know how to debug this. I can't even attach to the intermediate processes with gdb, since they're running in the flatpak container, and `flatpak enter` doesn't seem to work properly, and we don't have gdb in org.gnome.Platform so it wouldn't matter even if I could.

**Comment by sophie-h** [comment]
That sounds similar to #5501 on first glance

**Comment by mcatanzaro** [comment]
A notable difference is the WebKit subprocesses actually are quitting, though. Only flatpak-spawn itself fails to quit.

**Comment by sophie-h** [comment]
> A notable difference is the WebKit subprocesses actually are quitting, though. Only flatpak-spawn itself fails to quit.

How did you determine  that this is different? Is the webkit process not listed in `flatpak ps`?

If I see it correctly, it's exactly the same for Loupe. Did you see how I mentioned that the listed pid doesn't even exist?

What might be confusing about my issue is that I framed it around `FLATPAK_SPAWN_FLAGS_WATCH_BUS`. But there are basically two issues here: The sandbox should already have quit when the process in the sandbox terminated. And then second, at least when the app quits, the sandbox should be terminated anyways due to `--watch-bus`. But probably the issues are related.

**Comment by mcatanzaro** [comment]
Since you kill with `SIGKILL`, loupe has no chance to exit cleanly and cannot itself drop ownership of its bus name, right? The only way for that to happen would be if the message bus (dbus-broker or dbus-daemon) can notice the process is gone and do so on its own. But then again: I guess the message bus probably does do that? It probably just looks for a HUP event? I'm not sure how it works. Maybe this is irrelevant and it's the same bug after all.

In this bug report, the WebKit subprocesses are exiting cleanly.

**Comment by sophie-h** [comment]
> Since you kill with SIGKILL, loupe has no chance to exit cleanly and cannot itself drop ownership of its bus name, right?

I have no idea how the spawn portal works. This was just one way to reproduce the issue. The issue is appearing on regular basis for Loupe. And AFAIK it's not based on Loupe crashing.

![image](https://github.com/user-attachments/assets/72ec5eeb-f9c9-4f5d-946b-90771174abaa)

But even if the app crashes, there should be a way to ensure that the sandboxed are terminated as well imo.

**Comment by smcv** [comment]
One situation where the message bus would not notice a process exiting is if the D-Bus connection file descriptor had been inherited, transferred or otherwise leaked to another process, so that it was still being held open by another process: what it actually detects is the socket being closed, not the process exiting.

**Comment by mcatanzaro** [comment]
Surely that socket would use CLOEXEC though.

**Comment by mcatanzaro** [comment]
> Surely that socket would use CLOEXEC though.

Yes, I just did a deep dive. flatpak-spawn uses `g_bus_get_sync()` which, via a bunch of different layers, winds up creating the socket in `g_socket()` in gsocket.c. It sets `SOCK_CLOEXEC` if defined by the C library. So that's not going to be the problem.

**Comment by mcatanzaro** [comment]
Bug #5501 turned out to be an unrelated problem.

I think the problem here must be what I originally guessed:

> Inspecting flatpak-spawn.c, I think the only way this happens is if the flatpak portal doesn't emit the SpawnExited D-Bus signal. (Or if flatpak-spawn is somehow broken and doesn't receive it for some reason.) Since xdg-dbus-proxy is still running as well, I assume the flatpak portal has failed to notice the WebKit subprocess exiting.

Without a reproducer, it may be very hard to debug this, but possibly there is a race condition somewhere.

**Comment by sophie-h** [comment]
> Bug [#5501](https://github.com/flatpak/flatpak/issues/5501) turned out to be an unrelated problem.

I'm not sure how you can know that. The *reproducer* I constructed to make it possible to work on that problem turns out to be unrelated. So it's not actually a reproducer. The reproducer was not the issue I was experiencing with Loupe. See my comment in the issue:

> It looks like it's possible to also trigger the bug in other ways than `SIGKILL`. But that is not sure and unclear how.

**Comment by mcatanzaro** [comment]
> I'm not sure how you can know that. 

WebKit's child processes all know how to quit themselves when the parent process exits, so WebKit is actually fine if `--watch-bus` is broken. That shouldn't actually cause any problems.
